Fixes #187 - Inline edit components

tstrachota opened the pull request:
**This PR adds:**

- reactabular formatters that change table fields to be inline editable
- wrapper component that makes `FormControl` inline editable

**Storybook**:
https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html

More specifically:
- [Table With Inline Edit Row](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Row)
- [Table With Inline Edit Cells](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Cells)
- [Table With Inline Edit Column](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Column)
- ~~Inline Edit in Forms~~

**Additional comments**:

- I intentionally decided not to use `reactabular-edit` transformers. With the current implementation of tables in pf-react and blank cells that are used in `TablePfProvider` it turned out to be easier and more transparent to add our custom formatter.
- Note that inputs in table related storybooks use `onBlur` instead of `onChange`. Unfortunately `reactabular` always creates new rows on state change and prevents reconciliation. That was causing loss of focus on each key press.

tstrachota commented:
@waldenraines I agree. I didn't add it from the start because I wanted people to agree on the overall design and then I'd add more styling.
How about `pficon-edit` from http://www.patternfly.org/styles/icons/ ?

tstrachota commented:
Thanks @Rohoover I missed that. I'll add the missing visual features to the form inputs.

tstrachota commented:
Updated, link to storybook pages specific to this PR are in the [PR description](https://github.com/patternfly/patternfly-react/pull/240#issue-170431145).

jgiardino commented:
Thanks for adding these examples, @tstrachota!

@dabeng is also working on [inline edit in forms for patternfly core](https://github.com/patternfly/patternfly/issues/967). I'm thinking we should focus on getting the css finalized in core, then work on getting the pf-react inline edit updated to match. Let me know if you have any thoughts or concerns about that.

tstrachota commented:
Thanks for the heads up @jgiardino.

I understand that it's preferable to merge it into the pf core first, but we're going to need react inline edits quite soon and I'd like to avoid implementing it solely on our side. So I'd prefer to merge this PR independently of @dabeng's code and update it later if needed. I'm afraid that we could miss the timeframe in which we need this code if we're waiting for the core.

Since we need inline edits only in tables, I'm fine with splitting this PR into two - one for tables and one for forms. The latter could then wait for the changes in core.
What do you think?

jgiardino commented:
I agree, you shouldn't wait for things to be added to patternfly core if this is a high priority item. Since the other examples—inline edit in table rows and in table cells—don't exist in patternfly core yet either, then I'm ok with including the inline edit in forms example, too. It's very possible that all of these will need updates once the patterns are added to patternfly core. 

In looking at the UX, the [design documentation](http://www.patternfly.org/pattern-library/forms-and-controls/inline-edit/#inline-edit-in-a-table-view) includes the following statement:
> Row Selection: Only one row can be edited at a time. Edit puts an entire row into edit mode. The row in edit mode should be visually distinct from surrounding rows.

Can you update your example so that only one row displays as editable?  Although, I'm not seeing any statement on the expected behavior when one row is in edit mode, and the user then clicks the Edit button for another row. @maryclarke, @Rohoover or @mcarrano Do you know what the expected behavior is?

mcarrano commented:
Good question @jgiardino . If that happens, seems like there needs to be some exception handling to prompt the user whether they want to cancel the first edit. Or the other possibility is to disable other rows whenever any row is in edit mode. Thoughts?

tstrachota commented:
Good points. Disabling the other rows when one of them is in edit mode is quite elegant solution.

tstrachota commented:
I rebased and updated the inline edit row example so that it disables edit buttons on other rows.

jgiardino commented on sass/patternfly-react/_inline-edit.scss:
```diff
@@ -0,0 +1,82 @@
+  }
+}
+
+.inline-edit-buttons {
```
Can you add `pf-` to make this class name specific to patternfly?

jgiardino commented:
Thanks for making updates, @tstrachota! I'd like to hear what @mcarrano and @Rohoover think about the interaction now that it's updated. There is just one UX issue that I notice:

1. Navigate to the [table with inline edit row](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Row&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
2. Click the Edit icon in the first row
3. In the first column, change the value to "Z"
4. Switch focus to the next column

* **Expected Result:** The row remains in the first row position until I click the OK button.
* **Actual Result:** The row switches position to the last row.

I would consider this a bug and also a usability issue. It's a bug because now the layout doesn't display as expected. But it's a usability issue because you're moving contents based on changes before those changes are confirmed. This can make it really hard for the user to find the row in an actual application and make more changes or confirm the changes.

Aside from the UX issue above, there are also some accessibility issues that I noticed. It would be great if you can address these, but I also understand if this is out of scope for you. Let me know how you want to proceed with these and we can capture this in a separate issue. 

- **Keyboard accessibility** - Try using only the keyboard without the mouse to navigate elements in the table. 
  - In the table with inline edit row, I can access the Edit button, but then I'm not able to shift focus from one input field to the next using the Tab key or any other to navigate, and it isn't obvious to me from looking at the runtime html what's preventing that.
  - In the table with inline edit cell, the Edit button cannot receive focus. The other elements can receive focus and I'm able to access them with just a keyboard, but I'm not able to trigger the OK or Cancel actions with the Enter key. Usually any element that has a click event associated with it , like links or buttons, will automatically be triggered by hitting the Enter key too, but I don't think that's true for `onMouseUp`.
  - For enabling keyboard accessibility where buttons are involved, try using `<Button bsStyle="link">[Icon component]</Button>` and try using `onClick` instead of `onMouseUp` to see if you get the Enter key behavior automatically.
- **Screen reader accessibility** - I think the main issue is lack of text descriptions for form fields and icons. 
  - For icon buttons that have no text label, you can add `aria-label="Button Label"`. I'd suggest using "Edit", "Save" and "Cancel" as default labels, but I'm assuming that these should be editable.  
  - For input fields, you can use `aria-labelledby="[id of another element]"` to associate text with the input. In this case, I would associate the `<th>` elements with the input fields.
- Something I recommend you trying is to switch the [jsx-ally](https://github.com/patternfly/patternfly-react/blob/master/.eslintrc#L3) rules to "on" and review the issues that it describes. This might help you identify where some of these issues are coming from.

mcarrano commented:
First, I agree with @jgiardino that updating the value of the first column should not cause the row to change position. Would expect this to stay in place until the change is accepted.

The other thing that bothered me is the way the buttons show up below the row as it looks like they are actually applying to the row below. I went back to the design docs and while they do appear in a drop down tab like this, there is no button body that I think allows for better continuity with the highlighted row above. @kybaker, can you look at this and comment from a visual perspective? 

Other than those 2 items, I think this looks great.

kybaker commented:
@mcarrano The website example is low fidelity and in that context I think the icons outside of a button work better than they will as high fidelity. I think these controls need to be in a button. I also think they need more padding around them than we see in the code example.

jgiardino commented:
Thanks, @mcarrano and @kybaker! 

For the visual styling suggestions, it might be easier for me to open a PR with css updates after this one is merged. Let me know what you want to do, @tstrachota.

tstrachota commented:
Thank you for your detailed usability review, @jgiardino. I'm sorry for the late update - I somehow missed email notifications about your comments.

I rebased and updated the PR:
- Table rows being re-ordered during edit was caused by implementation of the storybook example. I fixed that.

- I added props for setting labels of confirm buttons in the components that I'm adding (`TableInlineEditHeaderRow`, `TableInlineEditRow`) with defaults "Save" and "Cancel". Users can easily localize them. The other buttons are all defined in the storybook examples. I can add labels there too just to present the best practices but it's always up to the users if they add them. In other words: it's just examples, not components that would be distributed with patternfly-react.

- Regarding the keyboard accessibility - I understand the importance but I do see it as a future improvement that is out of scope of this PR.

- I also added a ["column edit"](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Column) example based on @Rohoover's mockups for editing subscriptions in [Katello](https://theforeman.org/plugins/katello/) where we're planning to use it.

tstrachota commented:
Thanks for your feedback @leoluk 

>    in most cases, you wouldn't want to edit more than one item at the item - maybe it should auto-confirm any pending edits if a different one is selected?

This is just matter of configuration. The same components can be used in an auto-confirm manner.

>    my brain expected the current content to be selected - guess that's how most inline edits work?

I'm not sure I understand. Do you expect the text in the input to be selected when the edit starts?

>    can't press enter to confirm or cancel using escape

That should also be matter of configuration, but I'll make sure it works and update the examples.

leoluk commented:
> I'm not sure I understand. Do you expect the text in the input to be selected when the edit starts?

Yes, exactly. Straight-forward to add, but maybe it would be useful to have as a knob?

tstrachota commented:
Updated:

- The cell edit example so that it selects the input text when it's switched to the editing mode. Enter and escape should work there too now.
- The column edit example has a knob for editing the whole table.

There's plenty of ways how inline edits and reactabular tables can be configured. I'd like to avoid adding more in this PR. The main purpose is to add the common components and styles. Configuration of the behavior is on users.

tstrachota commented:
Ping. I rebased the PR. Any further comments about the components that I'm adding here?

waldenraines commented:
@tstrachota there are a couple of lint failures in the build.

tstrachota commented:
Ooops, updated, should be ok now.

waldenraines commented:
@tstrachota I guess you need to add > 2.7% more tests 😄

tstrachota commented:
I added some more tests, coveralls seems to be satisfied now :)

tstrachota commented:
@jgiardino whats the next steps, please? Could someone re-review it?

jeff-phillips-18 commented on package.json:
```diff
@@ -30,7 +30,7 @@
   "optionalDependencies": {
-    "lodash.orderby": "^4.6.0",
+    "lodash": "^4.17.0",
     "sortabular": "^1.5.1",
```
Is there a reason to pull in all of lodash? Looks like only orderby is being used here.

tstrachota replied:
I'm using also `cloneDeep` and `findIndex`.

jeff-phillips-18 replied:
OK, can we move this to devDependencies then?

tstrachota replied:
Sure, moved.

jeff-phillips-18 commented on sass/patternfly-react/_inline-edit.scss:
```diff
@@ -0,0 +1,127 @@
+  .spinner {
+    margin-left: 6px;
+  }
+}
```
Need to add the less equivalent as well (until #169 gets implemented).

tstrachota replied:
Done

jeff-phillips-18 commented on src/components/Form/InlineEditFormControl.js:
```diff
@@ -0,0 +1,137 @@
+    } else {
+      buttons = [
+        <ConfirmButton
+          bsStyle="primary"
```
The `primary` styling should be on the ConfirmButton so that any use outside of the `InlineEditFormControl` gets the proper styling.

jeff-phillips-18 commented on src/components/Table/Table.test.js:
```diff
@@ -134,3 +134,95 @@ test('Patternfly table renders properly', () => {
+      columns={mockPatternflyColumns}
+      components={options.components}
+    >
+      <Table.Header onRow={options.onHeaderRow || (() => ({}))} />
```
Could use `helpers.noop` here.

tstrachota commented:
Thanks for your review @jeff-phillips-18 
I addressed your comments. May I ask you to take another look?

tstrachota commented:
Rebased

jeff-phillips-18 commented on src/components/Table/TableConfirmButtonsRow.js:
```diff
@@ -0,0 +1,146 @@
+    this.saveRowDimensions = this.saveRowDimensions.bind(this);
+    this.handleScroll = this.handleScroll.bind(this);
+    this.handleResize = this.handleResize.bind(this);
+  }
```
We have standardized on using `helpers.bindMethods`

jeff-phillips-18 commented on src/components/Table/TableConfirmButtonsRow.js:
```diff
@@ -0,0 +1,146 @@
+    this.fetchClientDimensions();
+    window.addEventListener('scroll', this.handleScroll);
+    window.addEventListener('resize', this.handleResize);
+  }
```
These should use `helpers.debounce`

tstrachota replied:
Fixed, both functions are now debounced in the constructor.

jeff-phillips-18 commented on src/components/Table/TableConfirmButtonsRow.js:
```diff
@@ -0,0 +1,146 @@
+  /** Inject confirm buttons positions */
+  buttonsPosition: PropTypes.func,
+  /** Inject confirm buttons classes */
+  buttonsClass: PropTypes.func,
```
Typically we pass the additional classes rather than injecting defaults. Could this be string of classes to add rather than a callback function to inject the default classes?

jeff-phillips-18 commented on src/components/Table/TableConfirmButtonsRow.js:
```diff
@@ -0,0 +1,146 @@
+  }
+}
+
+TableConfirmButtonsRow.shouldComponentUpdate = true;
```
Is there a need for this?

tstrachota replied:
There is. See https://github.com/reactabular/reactabular/blob/master/docs/table/overriding-should-component-update.md and the reactabular code.
The default imlementation of `shouldComponentUpdate` allows updates only when columns or row data have changed. This is too strict in our case. We need to re-render components when the editing state is switched independently of row data change.

tstrachota commented:
@jeff-phillips-18 rebased and updated, could you take another look?

jgiardino commented:
I mentioned earlier in the PR that the inline edit form is currently being worked in patternfly core, and just wanted to update everyone that it's really close to being merged: https://github.com/patternfly/patternfly/pull/979

If this one is also close to being merged, we can create a separate issue to capture adopting the core css for this. Just let me know if that's needed. :-)

jeff-phillips-18 commented:
The core PR for inline form editing has merged, however the markup structure is quite different from what was done here and I don't believe its an easy update 😢 

I think it'd be best to separate the table inline editing and get that part in, then handle the inline form editing in a separate PR using the Core PF structure/styling.

tstrachota commented:
I don't care so much about the inline edit fields in forms. I can either adjust the styles later as you suggest or simply drop this part from my PR.

What Katello depends on and what we're going to need really soon is the tables part. If this doesn't get merged by Wednesday I'll have to copy the code to Katello - just temporarily I hope :(

jeff-phillips-18 commented:
OK, let's just address the table part in this PR and do the form part in a separate PR.

tstrachota commented:
Done, form related styles, components and storybook removed.

priley86 commented:
i'd like to review this some more this week from a consumption standpoint. we've already implemented some additional features around disabling row selection/select all that i'd like to review from a UX standpoint as well. It may also be preferable to allow inline edit in our use case (not positive yet). This has been on my backlog, but i'm going to try make time this week ;)

https://github.com/priley86/miq_v2v_ui_plugin/issues/207

jeff-phillips-18 commented:
@patternfly/patternfly-react-ux  Can we get a UX review for the inline table editing?

lizsurette commented:
@jeff-phillips-18 I think this looks awesome. I'm late to the game here and don't have all of the history, but testing it out, IMO it's ready to go. I think it would be worth an ACK from @Rohoover or @jgiardino though since they have been reviewing from the beginning.

jeff-phillips-18 commented:
@patternfly/patternfly-react-ux Should there be some spacing between the save and cancel buttons?

priley86 commented on src/components/InlineEdit/CancelButton.test.js:
```diff
@@ -0,0 +1,9 @@
+import { CancelButton } from './index';
+
+test('it renders properly', () => {
+  const component = render(<CancelButton />);
```
can we use `shallow` instead for all snapshots now (i.e. tests that are not interactive should not require `render`)? See WizardPattern...

tstrachota commented:
@jgiardino I fixed both the buttons position (nice catch btw) and sorting. It should behave consistently across the examples.

@Rohoover single column edit vs. full table edit is matter of configuration. There's a knob for that. I changed the default to be full-table.

tstrachota commented:
@priley86 I changed the buttons test to use `shallow`.

tstrachota commented:
@Rohoover @jeff-phillips-18 I added 4px space between the buttons.

jgiardino commented:
@tstrachota 
The updates to the [Table With Inline Edit Column](https://rawgit.com/tstrachota/patternfly-react/inline_edit_storybook/index.html?knob-Edit%20whole%20table=true&selectedKind=Table&selectedStory=Table%20With%20Inline%20Edit%20Column&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) look great!

Regarding @Rohoover's comment:
> I'm being a little picky here but can you make the checkbox the same size as the X? They are currently different widths.

I totally don't disagree with this from a design perspective. However, I expect details like this to be defined in patternfly core. If this is the expectation from a visual design perspective, then patternfly core should have an additional button class that supports this case. 

Does anyone object to filing an issue for this against patternfly-design, to let them dictate what the styling should be, and then to open an issue in patternfly core to provide the css? And to not create a one-off solution for this component? My thinking is that the width difference is really not noticeable by most users and therefore not worth the additional work to add then later update.

tstrachota commented:
@Rohoover I added a css line to fix the widths.

Since this PR includes styles for the inline edit elements that we'll most move to the core one day, we'll need to update this code anyway. Therefore I vote for sorting the equal width btn class asynchronously of this PR.

tstrachota commented:
Rebased

jgiardino commented:
I'm marking this as `css approved` since the remaining work that's needed is really minor and not worth blocking this. 

I created an issue in patternfly core to track the issue with the buttons: https://github.com/patternfly/patternfly/issues/1027

When that issue is resolved, we can revisit this component and make updates as needed.

In general, we shouldn't set a fixed pixel width for this type of element, since the size of the icon could change based on other factors, and therefore be larger than the fixed width of the button. Ideally, the width should be based on the font-size variable that controls the size of the icon. 

The reason why the icons here are inconsistent in width is due to one being a pf icon and one a font awesome icon, which are just different in width. In the pf core example for inline edit in a form, those icons are glyphicons, and appear to be implemented to be of consistent size (which is a nice feature and why this isn't an issue in that case).

priley86 commented on src/components/Table/TableConfirmButtonsRow.js:
```diff
@@ -0,0 +1,149 @@
+        createPortal(
+          this.renderConfirmButtons(),
+          this.props.buttonsMountpoint ||
+            this.element.closest('table').parentNode
```
do any of the downstream products consuming this care about IE11? If so - maybe we just include a [polyfill](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest) for element.closest.

priley86 commented on src/components/Table/__mocks__/mockInlineEditCellTable.js:
```diff
@@ -0,0 +1,661 @@
+                  inlineEditController.onChange(e.target.value, additionalData)
+                }
+                onKeyUp={e => {
+                  if (e.keyCode === 13) {
```
can you use helpers.KEY_CODES.ENTER_KEY ?

priley86 commented on src/components/Table/__mocks__/mockInlineEditCellTable.js:
```diff
@@ -0,0 +1,661 @@
+                      e.target.value,
+                      additionalData
+                    );
+                  } else if (e.keyCode === 27) {
```
can you use helpers.KEY_CODES.ESCAPE_KEY ?

priley86 commented:
i'm happy enough w/ this for the initial implementation. nice work! 👍

priley86 commented:
first commit needs to be semantic before merge...i.e. `feat(inline-edit): table inline edit components`

tstrachota commented:
@priley86 thanks, I added the key codes and polyfill for `closest`

tstrachota commented:
@jeff-phillips-18  oops, I just dismissed your review by pushing some tests for polyfill, I'm sorry.
